nistats/glm.py

```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
     -------
     Y : array of shape (n_time_points, n_voxels),
        the data after mean-scaling, de-meaning and multiplication by 100
+
     mean : array of shape (n_voxels,)
         the data mean
     """
     mean = Y.mean(0)
+    if (mean == 0).any():
+        warn('Mean values of 0 observed.'
+             'The data have probably been centered.'
+             'Scaling might not work as expected')
+    mean = np.maximum(mean, 1)
     Y = 100 * (Y / mean - 1)
     return Y, mean
 
 
-class GeneralLinearModel(object):
-    """ This class handles the so-called on General Linear Model
+def session_glm(Y, X, noise_model='ar1', bins=100):
+    """ GLM fit for an fMRI data matrix
 
-    Most of what it does in the fit() and contrast() methods
-    fit() performs the standard two-step ('ols' then 'ar1') GLM fitting
-    contrast() returns a contrast instance, yileding statistics and p-values.
-    The link between fit() and constrast is done vis the two class members:
+    Parameters
+    ----------
+    Y : array of shape(n_time_points, n_voxels)
+        the fMRI data
+
+    X : array of shape (n_time_points, n_regressors)
+           the design matrix
 
-    glm_results : dictionary of nistats.regression.RegressionResults instances,
-                 describing results of a GLM fit
+    noise_model : {'ar1', 'ols'}, optional
+        the temporal variance model. Defaults to 'ar1'
 
-    labels : array of shape(n_voxels),
-            labels that associate each voxel with a results key
+    bins : int, optional
+        Maximum number of discrete bins for the AR(1) coef histogram
+
+    Returns
+    -------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
+        keys correspond to the different labels values
+        values are RegressionResults instances corresponding to the voxels
     """
+    if noise_model not in ['ar1', 'ols']:
+        raise ValueError(' %s Unknown noise model' % noise_model)
 
-    def __init__(self, X):
-        """
-        Parameters
-        ----------
-        X : array of shape (n_time_points, n_regressors)
-           the design matrix
-        """
-        self.X = X
-        self.labels_ = None
-        self.results_ = None
+    if Y.shape[0] != X.shape[0]:
+        raise ValueError('Response and predictors are inconsistent')
```

Comment:
Maybe the message can be improved here too. Saying they don't have the number of rows is more precise and outputting the shape for both X and Y would help too.